File: app/src/main/java/com/jasonstudio/cookbook2/RecipeNutritionClasses/NutritionFragment.java
package com.jasonstudio.cookbook2.RecipeNutritionClasses;

import android.graphics.Color;
import android.os.Bundle;

import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.jasonstudio.cookbook2.R;
import com.github.mikephil.charting.charts.HorizontalBarChart;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.LegendEntry;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.formatter.IndexAxisValueFormatter;
import com.github.mikephil.charting.formatter.ValueFormatter;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;


/**
 * A simple {@link Fragment} subclass.
 * Use the {@link NutritionFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class NutritionFragment extends Fragment {
    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";

    // TODO: Rename and change types of parameters
    private String mParam1;
    private String mParam2;

    public NutritionFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @param param2 Parameter 2.
     * @return A new instance of fragment NutritionFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static NutritionFragment newInstance(String param1, String param2) {
        NutritionFragment fragment = new NutritionFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        args.putString(ARG_PARAM2, param2);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
            mParam2 = getArguments().getString(ARG_PARAM2);
        }
    }

    private String recipeId;
    private HorizontalBarChart mainChart;
    private HorizontalBarChart otherChart;
    private ArrayList<BarEntry> mainBarEntries;
    private ArrayList<BarEntry> otherBarEntries;
    private ArrayList<String> mainBarLabels;
    private ArrayList<String> otherBarLabels;
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_nutrition, container, false);
        mainChart = view.findViewById(R.id.main_nutrition_chart);
        otherChart = view.findViewById(R.id.other_nutrition_chart);
        mainChart.clear();
        otherChart.clear();
        chartParameterSetting(otherChart,Color.GREEN);
        chartParameterSetting(mainChart,Color.YELLOW);
        Bundle bundle = getArguments();
        recipeId = bundle.getString("recipe_id");
        loadNutritionData(recipeId);

        // Inflate the layout for this fragment
        return view;
    }

    private void loadNutritionData(String recipeId){
        RequestQueue queue = Volley.newRequestQueue(getContext());
        String url = "https://api.spoonacular.com/recipes/"+recipeId+"/nutritionWidget.json?apiKey="+getResources().getString(R.string.apiKeyUsing);
        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, url, null, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                try {
                    setChartData(response);
                    buildMainChart();
                    buildOtherChart();
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                error.printStackTrace();
            }
        });
        queue.add(request);
    }

    private void setChartData(JSONObject response) throws JSONException {
        mainBarEntries = new ArrayList<>();
        otherBarEntries = new ArrayList<>();
        mainBarLabels = new ArrayList<>();
        otherBarLabels = new ArrayList<>();
        JSONArray bad = response.getJSONArray("bad");
        for(int i=0; i<bad.length(); i++){
            JSONObject nutritionData = bad.getJSONObject(i);
            mainBarEntries.add(new BarEntry((float)i,(float)nutritionData.getDouble("percentOfDailyNeeds"),nutritionData.getString("amount")));
            mainBarLabels.add(nutritionData.getString("title"));
        }
        JSONArray good = response.getJSONArray("good");
        for(int i=0;i<good.length(); i++){
            JSONObject nutritionData = good.getJSONObject(i);
            otherBarEntries.add(new BarEntry((float)i,(float)nutritionData.getDouble("percentOfDailyNeeds"),nutritionData.getString("amount")));
            otherBarLabels.add(nutritionData.getString("title"));
        }
    }

    private void buildMainChart(){
        BarDataSet dataSet = new BarDataSet(mainBarEntries, "Tenses");
        dataSet.setColors(colorSet(mainBarEntries,Color.YELLOW));
        BarData data = new BarData(dataSet);
        data.setValueTextSize(9f);
        data.setValueTextColor(Color.WHITE);
        data.setBarWidth(0.2f);
        data.setValueFormatter(new ValueFormatter() {
            @Override
            public String getBarLabel(BarEntry barEntry) {
                return (Float.toString(barEntry.getY())+"% ("+(String)barEntry.getData()+")");
            }});
        mainChart.setData(data);
        mainChart.getXAxis().setValueFormatter(new IndexAxisValueFormatter(mainBarLabels));
        mainChart.getXAxis().setLabelCount(mainBarEntries.size());
        mainChart.getAxisLeft().setAxisMaximum(maxBarLength(mainBarEntries)+40f);
        chartParameterSetting(mainChart,Color.YELLOW);
    }

    private void buildOtherChart(){
        BarDataSet dataSet = new BarDataSet(otherBarEntries, "Tenses");
        dataSet.setColors(colorSet(otherBarEntries,Color.GREEN));
        BarData data = new BarData(dataSet);
        data.setValueTextSize(9f);
        data.setValueTextColor(Color.WHITE);
        data.setBarWidth(0.2f);
        data.setValueFormatter(new ValueFormatter() {
            @Override
            public String getBarLabel(BarEntry barEntry) {
                return (Float.toString(barEntry.getY())+"% ("+(String)barEntry.getData()+")");
            }});
        otherChart.setData(data);
        otherChart.getXAxis().setValueFormatter(new IndexAxisValueFormatter(otherBarLabels));
        otherChart.getXAxis().setLabelCount(otherBarEntries.size());
        otherChart.getAxisLeft().setAxisMaximum(maxBarLength(otherBarEntries)+40f);
        chartParameterSetting(otherChart,Color.GREEN);
    }

    private void chartParameterSetting(HorizontalBarChart chart,int color){

        chart.getXAxis().setTextColor(Color.WHITE);
        chart.getXAxis().setPosition(XAxis.XAxisPosition.BOTTOM);
        chart.getXAxis().setDrawGridLines(false);
        chart.animateY( 1000);
        chart.setDrawValueAboveBar(true);
        chart.getAxisLeft().setTextColor(Color.WHITE);
        chart.getAxisRight().setEnabled(false);
        chart.getDescription().setEnabled(false);
        chart.getLegend().setEnabled(true);
        ArrayList<LegendEntry> array = new ArrayList<>();
        for(int i =0; i<1; i++){
            array.add(new LegendEntry("% Of Daily Needs", Legend.LegendForm.DEFAULT,8f,8f,null,color));
        }
        chart.getLegend().setTextColor(color);
        chart.getLegend().setCustom(array);
        chart.setViewPortOffsets(200, 60, 120, 60);
        chart.getAxisLeft().setAxisMinimum(0);
        chart.invalidate();
    }

    private int[] colorSet(ArrayList<BarEntry> data, int color){
        float max = maxBarLength(data);

        int[] result = new int[data.size()];
        for(int i=0; i < data.size();i++){
            float value = data.get(i).getY();
            if(color == Color.GREEN){
                result[i] = Color.rgb((int)(160*(0.5*max+value)/(1.5*max)),(int)(255*(0.5*max+value)/(1.5*max)),(int)(50*(0.5*max+value)/(1.5*max)));
            }else if (color == Color.YELLOW){
                result[i] = Color.rgb((int)(255*(0.5*max+value)/(1.5*max)),(int)(215*(0.5*max+value)/(1.5*max)),(int)(30*(0.5*max+value)/(1.5*max)));
            }
        }
        return result;
    }

    private float maxBarLength(ArrayList<BarEntry> data){
        float max = 0;
        for(int i=0; i < data.size();i++){
            max = Math.max(max,data.get(i).getY());
        }
        return  max;
    }
}

File: app/src/main/java/com/jasonstudio/cookbook2/RecipeNutritionClasses/NutritionFragment.kt
package com.jasonstudio.cookbook2.RecipeNutritionClasses

import android.graphics.Color
import android.os.Bundle
import com.jasonstudio.cookbook2.RecipeNutritionClasses.NutritionFragment
import com.github.mikephil.charting.charts.HorizontalBarChart
import com.github.mikephil.charting.data.BarEntry
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.android.volley.Request
import com.jasonstudio.cookbook2.R
import com.android.volley.RequestQueue
import com.android.volley.toolbox.Volley
import com.android.volley.toolbox.JsonObjectRequest
import org.json.JSONObject
import org.json.JSONException
import com.android.volley.VolleyError
import com.github.mikephil.charting.components.Legend
import kotlin.Throws
import org.json.JSONArray
import com.github.mikephil.charting.data.BarDataSet
import com.github.mikephil.charting.data.BarData
import com.github.mikephil.charting.formatter.IndexAxisValueFormatter
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.components.LegendEntry
import com.github.mikephil.charting.formatter.ValueFormatter
import java.util.ArrayList

/**
 * A simple [Fragment] subclass.
 * Use the [NutritionFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class NutritionFragment : Fragment() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    private lateinit var recipeId: String
    private lateinit var mainChart: HorizontalBarChart
    private lateinit var otherChart: HorizontalBarChart
    private lateinit var mainBarEntries: ArrayList<BarEntry>
    private lateinit var otherBarEntries: ArrayList<BarEntry>
    private lateinit var mainBarLabels: ArrayList<String>
    private lateinit var otherBarLabels: ArrayList<String>
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_nutrition, container, false)
        mainChart = view.findViewById(R.id.main_nutrition_chart)
        otherChart = view.findViewById(R.id.other_nutrition_chart)
        mainChart.clear()
        otherChart.clear()
        chartParameterSetting(otherChart, Color.GREEN)
        chartParameterSetting(mainChart, Color.YELLOW)
        val bundle = arguments
        recipeId = bundle!!.getString("recipe_id") ?: ""
        loadNutritionData(recipeId)

        // Inflate the layout for this fragment
        return view
    }

    private fun loadNutritionData(recipeId: String?) {
        val queue = Volley.newRequestQueue(context)
        val url =
            "https://api.spoonacular.com/recipes/" + recipeId + "/nutritionWidget.json?apiKey=" + resources.getString(
                R.string.apiKeyUsing
            )
        val request = JsonObjectRequest(Request.Method.GET, url, null, { response ->
            try {
                setChartData(response)
                buildMainChart()
                buildOtherChart()
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }) { error -> error.printStackTrace() }
        queue.add(request)
    }

    @Throws(JSONException::class)
    private fun setChartData(response: JSONObject) {
        mainBarEntries = ArrayList()
        otherBarEntries = ArrayList()
        mainBarLabels = ArrayList()
        otherBarLabels = ArrayList()
        val bad = response.getJSONArray("bad")
        for (i in 0 until bad.length()) {
            val nutritionData = bad.getJSONObject(i)
            mainBarEntries!!.add(
                BarEntry(
                    i.toFloat(),
                    nutritionData.getDouble("percentOfDailyNeeds").toFloat(),
                    nutritionData.getString("amount")
                )
            )
            mainBarLabels!!.add(nutritionData.getString("title"))
        }
        val good = response.getJSONArray("good")
        for (i in 0 until good.length()) {
            val nutritionData = good.getJSONObject(i)
            otherBarEntries!!.add(
                BarEntry(
                    i.toFloat(),
                    nutritionData.getDouble("percentOfDailyNeeds").toFloat(),
                    nutritionData.getString("amount")
                )
            )
            otherBarLabels!!.add(nutritionData.getString("title"))
        }
    }

    private fun buildMainChart() {
        val dataSet = BarDataSet(mainBarEntries, "Tenses")
        dataSet.setColors(*colorSet(mainBarEntries, Color.YELLOW))
        val data = BarData(dataSet)
        data.setValueTextSize(9f)
        data.setValueTextColor(Color.WHITE)
        data.barWidth = 0.2f
        data.setValueFormatter(object : ValueFormatter() {
            override fun getBarLabel(barEntry: BarEntry): String {
                return java.lang.Float.toString(barEntry.y) + "% (" + barEntry.data as String + ")"
            }
        })
        mainChart!!.data = data
        mainChart!!.xAxis.valueFormatter = IndexAxisValueFormatter(mainBarLabels)
        mainChart!!.xAxis.labelCount = mainBarEntries!!.size
        mainChart!!.axisLeft.axisMaximum = maxBarLength(mainBarEntries) + 40f
        chartParameterSetting(mainChart, Color.YELLOW)
    }

    private fun buildOtherChart() {
        val dataSet = BarDataSet(otherBarEntries, "Tenses")
        dataSet.setColors(*colorSet(otherBarEntries, Color.GREEN))
        val data = BarData(dataSet)
        data.setValueTextSize(9f)
        data.setValueTextColor(Color.WHITE)
        data.barWidth = 0.2f
        data.setValueFormatter(object : ValueFormatter() {
            override fun getBarLabel(barEntry: BarEntry): String {
                return java.lang.Float.toString(barEntry.y) + "% (" + barEntry.data as String + ")"
            }
        })
        otherChart!!.data = data
        otherChart!!.xAxis.valueFormatter = IndexAxisValueFormatter(otherBarLabels)
        otherChart!!.xAxis.labelCount = otherBarEntries!!.size
        otherChart!!.axisLeft.axisMaximum = maxBarLength(otherBarEntries) + 40f
        chartParameterSetting(otherChart, Color.GREEN)
    }

    private fun chartParameterSetting(chart: HorizontalBarChart?, color: Int) {
        chart!!.xAxis.textColor = Color.WHITE
        chart.xAxis.position = XAxis.XAxisPosition.BOTTOM
        chart.xAxis.setDrawGridLines(false)
        chart.animateY(1000)
        chart.setDrawValueAboveBar(true)
        chart.axisLeft.textColor = Color.WHITE
        chart.axisRight.isEnabled = false
        chart.description.isEnabled = false
        chart.legend.isEnabled = true
        val array = ArrayList<LegendEntry>()
        for (i in 0..0) {
            array.add(
                LegendEntry(
                    "% Of Daily Needs",
                    Legend.LegendForm.DEFAULT,
                    8f,
                    8f,
                    null,
                    color
                )
            )
        }
        chart.legend.textColor = color
        chart.legend.setCustom(array)
        chart.setViewPortOffsets(200f, 60f, 120f, 60f)
        chart.axisLeft.axisMinimum = 0f
        chart.invalidate()
    }

    private fun colorSet(data: ArrayList<BarEntry>?, color: Int): IntArray {
        val max = maxBarLength(data)
        val result = IntArray(data!!.size)
        for (i in data.indices) {
            val value = data[i].y
            if (color == Color.GREEN) {
                result[i] = Color.rgb(
                    (160 * (0.5 * max + value) / (1.5 * max)).toInt(),
                    (255 * (0.5 * max + value) / (1.5 * max)).toInt(),
                    (50 * (0.5 * max + value) / (1.5 * max)).toInt()
                )
            } else if (color == Color.YELLOW) {
                result[i] = Color.rgb(
                    (255 * (0.5 * max + value) / (1.5 * max)).toInt(),
                    (215 * (0.5 * max + value) / (1.5 * max)).toInt(),
                    (30 * (0.5 * max + value) / (1.5 * max)).toInt()
                )
            }
        }
        return result
    }

    private fun maxBarLength(data: ArrayList<BarEntry>?): Float {
        var max = 0f
        for (i in data!!.indices) {
            max = Math.max(max, data[i].y)
        }
        return max
    }

    companion object {
        // TODO: Rename parameter arguments, choose names that match
        // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
        private const val ARG_PARAM1 = "param1"
        private const val ARG_PARAM2 = "param2"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param param1 Parameter 1.
         * @param param2 Parameter 2.
         * @return A new instance of fragment NutritionFragment.
         */
        // TODO: Rename and change types and number of parameters
        fun newInstance(param1: String?, param2: String?): NutritionFragment {
            val fragment = NutritionFragment()
            val args = Bundle()
            args.putString(ARG_PARAM1, param1)
            args.putString(ARG_PARAM2, param2)
            fragment.arguments = args
            return fragment
        }
    }
}